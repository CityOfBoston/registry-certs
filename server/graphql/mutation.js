// @flow

import type { Context } from '.';

import { calculateCost } from '../../lib/costs';

export const Schema = `
input CertificateOrderItem {
  id: String!
  name: String!
  quantity: Int!
}

type SubmittedOrder {
  id: String!
}

type Mutation {
  submitDeathCertificateOrder (
    contactName: String!
    contactEmail: String!
    contactPhone: String!

    shippingName: String!
    shippingCompanyName: String
    shippingAddress1: String!
    shippingAddress2: String
    shippingCity: String!
    shippingState: String!
    shippingZip: String!

    cardToken: String!
    cardLast4: String!

    cardholderName: String!
    billingAddress1: String!
    billingAddress2: String
    billingCity: String!
    billingState: String!
    billingZip: String!

    items: [CertificateOrderItem!]!
  ): SubmittedOrder!
}
`;

type CertificateOrderItem = {|
  id: string,
  name: string,
  quantity: number,
|};

type SubmittedOrder = {|
  id: string,
|};

type submitDeathCertificateOrderArgs = {|
  contactName: string,
  contactEmail: string,
  contactPhone: string,

  shippingName: string,
  shippingCompanyName: string,
  shippingAddress1: string,
  shippingAddress2: string,
  shippingCity: string,
  shippingState: string,
  shippingZip: string,

  cardToken: string,
  cardLast4: string,

  cardholderName: string,
  billingAddress1: string,
  billingAddress2: string,
  billingCity: string,
  billingState: string,
  billingZip: string,

  items: Array<CertificateOrderItem>,
|};

export const resolvers = {
  Mutation: {
    submitDeathCertificateOrder: async (
      root: mixed,
      args: submitDeathCertificateOrderArgs,
      { stripe }: Context
    ): Promise<SubmittedOrder> => {
      const { cardToken, items } = args;

      const totalQuantity = items.reduce(
        (count, { quantity }) => count + Math.max(quantity, 0),
        0
      );

      if (totalQuantity === 0) {
        throw new Error('Quantity of order is 0');
      }

      const { total } = calculateCost(totalQuantity);

      const charge = await stripe.charges.create({
        amount: total,
        currency: 'usd',
        source: cardToken,
        description: 'Death certificates (Registry)',
      });

      return { id: charge.id };
    },
  },
};
