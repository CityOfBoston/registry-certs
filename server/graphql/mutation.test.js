// @flow

import { resolvers } from './mutation';

const DEFAULT_ORDER = {
  contactName: '',
  contactEmail: '',
  contactPhone: '',

  shippingName: '',
  shippingCompanyName: '',
  shippingAddress1: '',
  shippingAddress2: '',
  shippingCity: '',
  shippingState: '',
  shippingZip: '',

  cardholderName: '',
  cardToken: 'tok_test',
  cardLast4: '1234',

  billingAddress1: '',
  billingAddress2: '',
  billingCity: '',
  billingState: '',
  billingZip: '',

  items: [
    {
      id: '12345',
      name: '',
      quantity: 10,
    },
  ],
};

describe('Mutation resolvers', () => {
  describe('submitDeathCertificateOrder', () => {
    it('throws if there are no items', async () => {
      await expect(
        resolvers.Mutation.submitDeathCertificateOrder(
          null,
          {
            ...DEFAULT_ORDER,
            items: [],
          },
          ({}: any)
        )
      ).rejects.toMatchSnapshot();
    });

    it('sends a charge to Stripe', async () => {
      const chargesCreate = jest.fn();
      const stripe: any = {
        charges: {
          create: chargesCreate,
        },
      };

      chargesCreate.mockReturnValue(Promise.resolve({ id: 'ch_12345' }));

      await resolvers.Mutation.submitDeathCertificateOrder(
        null,
        DEFAULT_ORDER,
        ({ stripe }: any)
      );

      expect(chargesCreate).toHaveBeenCalledWith({
        amount: 14449,
        currency: 'usd',
        source: 'tok_test',
        description: 'Death certificates (Registry)',
      });
    });
  });
});
