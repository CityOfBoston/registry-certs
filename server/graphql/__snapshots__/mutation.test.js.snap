// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation resolvers submitDeathCertificateOrder throws if there are no items 1`] = `[Error: Quantity of order is 0]`;
